### PR TITLE
feat: make message input w data persist when covering fees

### LINKED ISSUE #1059: message input w data swallowed when covering fees
The process of finding resources to cover tx fees removes current base asset inputs (coins and messages) and replaces them with equivalent inputs that can cover both the tx fee and the original base asset that was to be transferred in the tx.

The current implementation may remove messages that have data in them. This should be avoided as the contract/script might have used the data in some way.

### PR DESCRIPTION
**Pull request by MujkicA**
Closes  #1059 

I've extended the filtering conditions when replacing existing base inputs in `add_fee_coins` to not remove messages with data from the existing input set. 

I explored some options to make the interface for resources/coins less confusing but I think it's better addressed in a separate PR once #1069 lands.

### Checklist
- [x] I have linked to any relevant issues.
- [ ] I have updated the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary labels.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: packages/fuels-core/src/types/wrappers/input.rs [hunk 1]
```diff
@@ -65,6 +65,18 @@ impl Input {
...
+    pub fn contains_data(&self) -> bool {
+        match self {
+            Self::ResourceSigned { resource, .. } | Self::ResourcePredicate { resource, .. } => {
+                match resource {
+                    CoinType::Message(msg) => !msg.data.is_empty(),
+                    _ => false,
+                }
+            },
+            _ => false,
+        }
```
Comment (segfault-magnet) [inline packages/fuels-core/src/types/wrappers/input.rs:77]:
```suggestion
        match self {
            Self::ResourceSigned {
                resource: CoinType::Message(msg),
                ..
            }
            | Self::ResourcePredicate {
                resource: CoinType::Message(msg),
                ..
            } => !msg.data.is_empty(),
            _ => false,
        }
```

A bit flatter with more pattern matching if you like it.

File: packages/fuels/tests/predicates.rs [hunk 2]
```diff
@@ -658,3 +675,40 @@ async fn predicate_configurables() -> Result<()> {
...
+    let coins = setup_single_asset_coins(
+        predicate.address(),
+        BASE_ASSET_ID,
+        1,
+        amount,
+    );
+    let message = get_test_message_w_data(predicate.address(), amount, Default::default());
+    let message_input = Input::resource_predicate(CoinType::Message(message.clone()), predicate.code().clone(), predicate.data().clone());
+
+    let (provider, _address) = setup_test_provider(coins, vec![message.clone()], None, None).await;
```
Comment (segfault-magnet) [inline packages/fuels/tests/predicates.rs:702]:
```suggestion
    let (provider, _) = setup_test_provider(coins, vec![message.clone()], None, None).await;
```

**Review by segfault-magnet — DISMISSED** [review]

**Review by hal3e — APPROVED** [review]
LGTM

**Review by segfault-magnet — APPROVED** [review]

**Review by Salka1988 — APPROVED** [review]
👍

**Pull request merged**